### PR TITLE
Add HTTP/2 support via nomadlabsinc/ht2 shard

### LINKED ISSUE #1966: Add support for HTTP2 to Lucky HTTP2 -- action plan
# HTTP/2 Support for Lucky Framework - Checklist

This document outlines the necessary changes to add HTTP/2 support to the Lucky framework. It's divided into creating a new Crystal shard for HTTP/2 handling and integrating this shard into Lucky.

# Crystal Tooling: `nomadlabsinc/ht2` Shard

This section covers the creation of a new Crystal shard, `nomadlabsinc/ht2`, designed to replace `Crystal::HTTP::Server` for HTTP/2 communication, and documentation for its usage.

### 1.1. Project Setup & Initialization
- [x] Initialize new shard: `crystal init shard ht2`.
- [x] Define `shard.yml` with necessary metadata (version, authors, license, description).
- [x] Add development dependencies (e.g., `crystal-db` if examples need it, testing frameworks).
- [x] Create basic directory structure (`src/ht2/`, `src/ht2/frame/`, `spec/`).
- [x] Define main module `HT2` and `HT2::Server` class structure.

#### 1.2.1. Connection Management
- [x] Implement TLS ALPN negotiation for "h2" protocol.
- [x] Implement reception and validation of client connection preface (`PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n`).
- [x] Implement sending of server connection preface (initial SETTINGS frame).
- [x] Manage connection state (e.g., active, going_away, closed).
- [x] Handle connection-level errors and send GOAWAY frames.
- [x] Implement PING frame handling for RTT measurement and keep-alive.

#### 1.2.2. Frame (De)serialization
- [x] Define a generic frame structure (length, type, flags, stream_id, payload).
- [x] Implement parsing for generic frame headers.
- [x] For **DATA** frames:
    - [x] Implement DATA frame parsing (padded, end_stream flag).
    - [x] Implement DATA frame serialization.
- [x] For **HEADERS** frames:
    - [x] Implement HEADERS frame parsing (padded, end_stream, end_headers, priority flags/fields).
    - [x] Implement HEADERS frame serialization.
- [x] For **PRIORITY** frames:
    - [x] Implement PRIORITY frame parsing (stream_dependency, weight).
    - [x] Implement PRIORITY frame serialization.
- [x] For **RST_STREAM** frames:
    - [x] Implement RST_STREAM frame parsing (error_code).
    - [x] Implement RST_STREAM frame serialization.
- [x] For **SETTINGS** frames:
    - [x] Implement SETTINGS frame parsing (ack flag, identifier-value pairs).
    - [x] Implement SETTINGS frame serialization.
    - [x] Handle SETTINGS synchronization and ACK.
    - [x] Define and manage settings parameters (e.g., `SETTINGS_HEADER_TABLE_SIZE`, `SETTINGS_ENABLE_PUSH`, `SETTINGS_MAX_CONCURRENT_STREAMS`, `SETTINGS_INITIAL_WINDOW_SIZE`, `SETTINGS_MAX_FRAME_SIZE`, `SETTINGS_MAX_HEADER_LIST_SIZE`).
- [x] For **PUSH_PROMISE** frames:
    - [x] Implement PUSH_PROMISE frame parsing (padded, end_headers, promised_stream_id, header_block_fragment).
    - [x] Implement PUSH_PROMISE frame serialization.
    - [x] (Optional) Implement server push logic if the server is to initiate pushes.
- [x] For **GOAWAY** frames:
    - [x] Implement GOAWAY frame parsing (last_stream_id, error_code, opaque_data).
    - [x] Implement GOAWAY frame serialization.
- [x] For **WINDOW_UPDATE** frames:
    - [x] Implement WINDOW_UPDATE frame parsing (window_size_increment).
    - [x] Implement WINDOW_UPDATE frame serialization.
- [x] For **CONTINUATION** frames:
    - [x] Implement CONTINUATION frame parsing (end_headers flag, header_block_fragment).
    - [x] Implement CONTINUATION frame serialization.
    - [x] Handle fragmentation of header blocks across HEADERS/PUSH_PROMISE and CONTINUATION frames.

#### 1.2.3. Stream Management
- [x] Implement stream lifecycle and state machine (idle, reserved (local/remote), open, half-closed (local/remote), closed).
- [x] Manage stream concurrency (`SETTINGS_MAX_CONCURRENT_STREAMS`).
- [x] Implement stream ID allocation (client-initiated odd, server-initiated even).
- [x] Handle stream-level errors and send RST_STREAM frames.
- [x] Implement stream priority management (dependency and weight).

#### 1.2.4. Header Compression (HPACK)
- [x] Integrate or implement HPACK encoder.
- [x] Integrate or implement HPACK decoder.
- [x] Manage dynamic table size (`SETTINGS_HEADER_TABLE_SIZE`).
- [x] Handle pseudo-headers (`:method`, `:scheme`, `:authority`, `:path`, `:status`) correctly.
- [x] Ensure all regular header field names are lowercase.

#### 1.2.5. Flow Control
- [x] Implement connection-level flow control using WINDOW_UPDATE.
- [x] Implement stream-level flow control using WINDOW_UPDATE.
- [x] Manage initial window sizes (`SETTINGS_INITIAL_WINDOW_SIZE`).
- [x] Handle window size updates and prevent negative window sizes.

### 1.3. API Design for `HT2::Server`
- [x] Design `HT2::Server.new(host : String, port : Int32, handler : Handler, tls_context : OpenSSL::SSL::Context::Server?)` constructor.
    - [x] Consider alternative handler types: `Proc(HT2::Request, HT2::Response, Nil)` handler.
- [x] Design `HT2::Server#listen` method (blocking, starts accepting connections).
- [x] Design `HT2::Server#close` method (graceful shutdown, GOAWAY).
- [x] Define `HT2::Request` object:
    - [x] Properties: `method`, `path`, `scheme`, `authority`, `headers`, `body_io`, `stream_id`, `version` (fixed to "HTTP/2").
    - [x] Access to HTTP/2 specific features (e.g., priority).
- [x] Define `HT2::Response` object:
    - [x] Properties: `status`, `headers`, `write` methods.
    - [x] Methods for setting headers and body.
    - [x] (Optional) Methods for server push: `push` method.
- [x] Define `HT2::Context` to encapsulate `Request`, `Response` for a stream.
    - [x] Pass `Request` and `Response` to the handler.
- [x] Design configuration options for the server (e.g., max_concurrent_streams, initial_window_size, max_frame_size, read_timeout, write_timeout).

### 1.4. Testing
- [x] Develop unit tests for frame (de)serialization for all frame types.
- [x] Develop unit tests for HPACK implementation.
- [x] Develop unit tests for stream state transitions and management.
- [x] Develop unit tests for flow control logic.
- [x] Develop integration tests for basic request/response cycles.
- [x] Develop integration tests for concurrent streams.
- [x] Develop integration tests for error handling (stream and connection errors).
- [x] (Optional) Test against existing HTTP/2 clients/tools (e.g., nghttp2, curl).

### 1.5. Documentation for `nomadlabsinc/ht2` Shard
- [x] Write a comprehensive `README.md`:
    - [x] Shard purpose and features.
    - [x] Installation instructions (`shard.yml` entry).
    - [x] Basic usage examples (setting up a simple server).
    - [x] Advanced usage examples (custom configurations, server push if implemented).
    - [x] API overview.
- [x] Add inline API documentation (Crystal `doc` comments) for all public types and methods.
- [x] Generate HTML API documentation using `crystal docs`.
- [x] Provide a guide on TLS setup for HTTP/2 (ALPN requirement).
- [x] Explain key HTTP/2 concepts relevant to users of the shard.

## 2. Integration Documentation for `nomadlabsinc/ht2` Consumers (e.g., Lucky Framework)

- [x] How to add `nomadlabsinc/ht2` as a dependency.
- [x] Step-by-step guide to replace `HTTP::Server` with `HT2::Server`.
    - [x] Instantiation differences.
    - [x] Handler signature adaptation.
- [x] Configuration mapping: How to set `HT2::Server` options (e.g., TLS context, port, host, HTTP/2 specific params).
- [x] Guidance on adapting `HTTP::Request` and `HTTP::Response` usage if the new shard's API differs significantly.
- [x] Best practices for using the `nomadlabsinc/ht2` shard.
- [x] How to leverage HTTP/2 specific features exposed by the shard (e.g., server push API).
- [x] Troubleshooting common issues.

#### Security Features
- [x] Implement protection against CVE-2023-44487 (Rapid Reset Attack).
- [x] Implement protection against settings floods.
- [x] Implement protection against ping floods.
- [x] Implement protection against priority floods.
- [x] Implement protection against HPACK bombs.
- [x] Implement header size violation detection.
- [x] Implement stream limit violation detection.
- [x] Implement frame size violation detection.

#### Performance Optimizations
- [x] Implement adaptive buffer management for dynamic buffer sizing.
- [x] Implement adaptive flow control with auto-tuning.
- [x] Implement backpressure management for write pressure.
- [x] Implement buffer pooling for buffer recycling.
- [x] Implement frame caching for pre-serialized frames.
- [x] Implement multi-frame writer for batched frame writing.
- [x] Implement vectored I/O for system-level optimization.
- [x] Implement zero-copy data forwarding.

#### HTTP/2 Clear Text (h2c) Support
- [x] Implement HTTP/1.1 Upgrade mechanism (RFC 7540 Section 3.2).
- [x] Implement Prior Knowledge support (RFC 7540 Section 3.4).
- [x] Implement automatic connection type detection.
- [x] Implement settings negotiation via HTTP2-Settings header.

#### Monitoring & Observability
- [x] Implement connection metrics tracking.
- [x] Implement performance metrics (latency/throughput).
- [x] Implement stream lifecycle tracing.
- [x] Implement frame-level debugging mode.

#### Framework Integration Examples
- [x] Create Kemal framework integration example.
- [x] Create standalone server example.
- [x] Create HTTP/2 Clear Text demo example.
- [x] Create Prior Knowledge demo example.

# Lucky Framework Changes

This section assumes `nomadlabsinc/ht2` shard is complete and available. It details changes required within the Lucky framework itself.

### 1.1. Configuration
- [ ] Add new configuration options in `config/server.cr` (or equivalent) for HTTP/2:
    - [ ] `config.http2.enabled = true/false` (default to `false` initially).
    - [ ] `config.http2.port` (if different from HTTP/1.1, though typically same port with ALPN).
    - [ ] `config.http2.tls_key_file` and `config.http2.tls_cert_file` (emphasize TLS requirement).
    - [ ] Expose relevant `HT2::Server` settings (e.g., `max_concurrent_streams`, `initial_window_size`).
- [ ] Update `lucky watch` and server startup commands (`Lucky::ServerRunner` or similar) to read and apply these configurations.
- [ ] Ensure development mode (auto-reload) is compatible with the new server setup.

### 1.2. Server Abstraction Layer (Recommended)
- [ ] Define a `Lucky::Server::AdapterInterface` (e.g., using an abstract class or module).
    - [ ] Common methods: `initialize(config)`, `listen`, `close`.
    - [ ] Common handler signature or adaptation logic.
- [ ] Create `Lucky::Server::HTTP1Adapter` wrapping `Crystal::HTTP::Server` and conforming to `AdapterInterface`.
    - [ ] This adapter would use existing Lucky HTTP/1.1 logic.
- [ ] Create `Lucky::Server::HTTP2Adapter` wrapping `HT2::Server` and conforming to `AdapterInterface`.
    - [ ] This adapter will integrate the new `nomadlabsinc/ht2` shard.
- [ ] Refactor `Lucky::ServerRunner` (or equivalent) to instantiate and use the appropriate adapter based on configuration (`config.http2.enabled`).

### 1.3. Request/Response Object Adaptation
- [ ] Review `Lucky::Request` and `Lucky::Response`.
- [ ] If `nomadlabsinc/ht2` uses its own request/response objects (`HT2::Request`, `HT2::Response`):
    - [ ] Implement mapping logic from `HT2::Request` to `Lucky::Request`.
        - [ ] Ensure correct population of `method`, `path`, `headers`, `body`, `params`, `cookies`, `remote_ip`, `scheme`, `host`, `port`.
        - [ ] Pay special attention to HTTP/2 pseudo-headers and lowercase header names.
    - [ ] Implement mapping logic from `Lucky::Response` to `HT2::Response`.
        - [ ] Ensure correct setting of `status_code`, `headers`, `body`.
- [ ] If `nomadlabsinc/ht2` can work directly with `HTTP::Request` and `HTTP::Response` (or compatible types), ensure full compatibility.
- [ ] Update `Lucky::Route` and dispatching mechanism to handle requests from the HTTP/2 server.

### 1.4. Middleware and Handler Compatibility
- [ ] Review all built-in Lucky middleware/handlers:
    - [ ] `Lucky::StaticFileHandler`: Consider opportunities for server push (e.g., pushing linked CSS/JS for an HTML file).
    - [ ] `Lucky::ForceSSLHandler`: Behavior might need adjustment as HTTP/2 is typically run over TLS. Could be a no-op or verify scheme.
    - [ ] `Lucky::ErrorHandler`: Ensure error responses are correctly formatted for HTTP/2.
    - [ ] `Lucky::Session::SaveSessionHandler`: Ensure cookie handling is compatible.
    - [ ] `Lucky::CSRFHandler`: Verify token mechanisms.
- [ ] Ensure all custom headers set by Lucky or application code are handled correctly (e.g. case-insensitivity for lookup, lowercase for sending).
- [ ] Update any middleware that directly interacts with `HTTP::Server::Context` to use the new `HT2::Context` or an abstracted version.

### 1.5. TLS Management
- [ ] Ensure Lucky's TLS configuration (`config.ssl_key_file`, `config.ssl_cert_file`) can be used by `HT2::Server`.
- [ ] Verify ALPN is correctly configured when TLS is enabled for HTTP/2.
- [ ] Update documentation regarding TLS setup, emphasizing its necessity for browser-based HTTP/2.

### 1.6. Testing
- [ ] Update existing integration and acceptance tests to run over an HTTP/2-enabled server.
- [ ] Add new tests specifically for HTTP/2 functionality if Lucky exposes any (e.g., server push APIs).
- [ ] Test behavior with mixed HTTP/1.1 and HTTP/2 configurations if supported.
- [ ] Perform thorough testing with various HTTP/2 clients (browsers, curl with HTTP/2).

### 1.7. Documentation Updates
- [ ] Update Lucky guides on server configuration to include HTTP/2 settings.
- [ ] Document any changes in behavior or new features related to HTTP/2.
- [ ] Provide guidance for application developers on how to leverage HTTP/2 (e.g., implications for asset loading, server push if available).
- [ ] Update deployment guides with considerations for HTTP/2 (e.g., reverse proxy configuration like Nginx/HAProxy for HTTP/2 termination or pass-through).

## 2. Integrating `nomadlabsinc/ht2` Shard - Multi-Step Action Plan

This assumes the Server Abstraction Layer (1.2) is being implemented.

### 2.1. Phase 1: Initial Setup and Basic Integration
- [ ] Add `nomadlabsinc/ht2` shard to Lucky's `shard.yml`.
- [ ] Implement the `Lucky::Server::AdapterInterface`.
- [ ] Implement `Lucky::Server::HTTP1Adapter` by refactoring existing `HTTP::Server` usage into it. Ensure Lucky works as before with this adapter.
- [ ] Implement the basic structure of `Lucky::Server::HTTP2Adapter`, instantiating `HT2::Server`.
- [ ] Modify `Lucky::ServerRunner` to select and use `HTTP1Adapter` or `HTTP2Adapter` based on `config.http2.enabled`.

### 2.2. Phase 2: Request/Response Path Adaptation
- [ ] Implement the handler logic within `Lucky::Server::HTTP2Adapter` to:
    - [ ] Receive `HT2::Request` and `HT2::Response` from `HT2::Server`.
    - [ ] Convert/map `HT2::Request` to `Lucky::Request`.
    - [ ] Pass `Lucky::Request` through Lucky's routing and action dispatch.
    - [ ] Receive `Lucky::Response` from the action.
    - [ ] Convert/map `Lucky::Response` back to `HT2::Response` and send it.
- [ ] Thoroughly test this request/response flow with simple routes and actions.

### 2.3. Phase 3: Middleware and Advanced Feature Compatibility
- [ ] Systematically review and update each piece of Lucky middleware (as per 1.4) for compatibility with the `HTTP2Adapter` flow.
- [ ] Test file serving, error handling, sessions, CSRF, etc., over HTTP/2.
- [ ] If `nomadlabsinc/ht2` supports server push and Lucky wants to expose it:
    - [ ] Design an API in `Lucky::Response` (e.g., `response.push("/assets/style.css")`).
    - [ ] Implement this API to call the underlying `HT2::Response#push` method.

### 2.4. Phase 4: Configuration, TLS, and Edge Cases
- [ ] Fully implement configuration loading for all `config.http2.*` settings in `Lucky::Server::HTTP2Adapter`.
- [ ] Ensure TLS context setup and ALPN negotiation are robust.
- [ ] Test various edge cases: large uploads/downloads, many concurrent requests, slow clients, error conditions (stream resets, connection drops).

### 2.5. Phase 5: Testing, Documentation, and Release Preparation
- [ ] Conduct comprehensive end-to-end testing of Lucky applications running with `config.http2.enabled = true`.
- [ ] Perform performance benchmarks comparing HTTP/1.1 (via `HTTP1Adapter`) and HTTP/2 (via `HTTP2Adapter`).
- [ ] Write all necessary documentation updates for Lucky users (guides, API docs).
- [ ] Prepare release notes detailing the new HTTP/2 support.
- [ ] Consider a beta release or feature flag period for wider testing.

---
Issue co-authored w/use of Gemini 2.5. Pro

This checklist provides a detailed roadmap. Each item may represent significant effort and further sub-tasks.

### PR DESCRIPTION
**Pull request by robcole**
## Summary

This PR resolves #1966 by adding HTTP/2 support to Lucky applications using the `nomadlabsinc/ht2` shard.

Key features:
- Conditional compilation support - applications can be compiled with either HTTP/1.1 or HTTP/2, but not both
- New `Lucky::HTTP2AppServer` abstract class for HTTP/2 servers
- HTTP/2 response adapter that bridges Lucky's interface with HT2::Response
- Comprehensive configuration options for HTTP/2 settings
- Complete upgrade documentation

## Changes Made

- **shard.yml**: Added ht2 dependency
- **src/lucky/server.cr**: Added HTTP/2 configuration settings (cert files, max streams, frame size)
- **src/lucky/http2_app_server.cr**: New abstract class for HTTP/2 app servers
- **src/lucky/http2_response_adapter.cr**: Response adapter bridging Lucky and HT2 interfaces
- **src/lucky.cr**: Conditional compilation support for HTTP/2 files
- **HTTP2_UPGRADE_STEPS.md**: Comprehensive upgrade guide for existing applications

## Usage

Applications can enable HTTP/2 by:
1. Installing the ht2 dependency
2. Configuring TLS certificates
3. Updating server class to inherit from `Lucky::HTTP2AppServer`
4. Compiling with `-Dhttp2` flag

## Test Plan

- ✅ Code compiles successfully without HTTP/2 flag (HTTP/1.1 only)
- ✅ Code compiles successfully with HTTP/2 flag
- ✅ Linting passes with proper formatting
- ✅ Follows Lucky's architectural patterns and conventions
- ✅ Comprehensive documentation provided

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by robcole** [comment]
@jwoertink -- Very much a WIP, just coming back from a week off w/family and wanted to get something up that felt a bit more "Lucky." Will be testing it out over the next few days but gimme a holler if there's any direction you'd like to give prior to that -- happy to chat it through in Discord.

File: src/lucky/http2_app_server.cr [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# HTTP2 App Server for Lucky applications
+#
+# This class provides HTTP/2 support using the ht2 shard.
+# It requires TLS certificates and is enabled via the http2_enabled setting.
+{% if flag?(:http2) %}
```
Comment (jwoertink) [inline src/lucky/http2_app_server.cr:5]: I'm not sure if we have access to pass this flag 🤔  When you run `lucky dev`, that tells the CLI to call Nox which runs the `Procfile.dev` that calls `lucky watch` that internally calls `crystal build`

https://github.com/luckyframework/lucky/blob/ff904b28c7a0fdcd45148be3850cc00d5984bb59/tasks/watch.cr#L331

I'd have to test, but I'm not sure if doing `lucky dev -Dhttp2` would ever get picked up by that build.

File: HTTP2_UPGRADE_STEPS.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  # HTTP/2 with TLS Configuration
+  settings.http2_enabled = true
+  settings.http2_cert_file = "./server.crt"      # Required for TLS
+  settings.http2_key_file = "./server.key"       # Required for TLS
+  settings.http2_max_concurrent_streams = 100    # Optional
+  settings.http2_max_frame_size = 16384          # Optional
+end
+```
+
+### Option B: HTTP/2 Cleartext (h2c) - for apps behind load balancers
```
Comment (jwoertink) [inline HTTP2_UPGRADE_STEPS.md:57]: Nice!

File: HTTP2_UPGRADE_STEPS.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  def listen
+    @server.bind_tcp(host, port)
+    @server.listen
+  end
+end
+```
+
+### After (HTTP/2):
+```crystal
+class AppServer < Lucky::HTTP2AppServer
```
Comment (jwoertink) [inline HTTP2_UPGRADE_STEPS.md:99]: Love how this is opt-in like this 🙌

**Review by jwoertink — COMMENTED** [review]
I have no clue if the actual logic is correct or not, so I can only assume... but the overall integration seems like the right direction. My only concern is that compile-time flag. I totally get it, I'm just not sure if it'll actually work. Can you think of any way we can check? Or maybe they're not even necessary? I mean, if you don't switch your base server then none of this will matter, right?

**Comment by robcole** [comment]
> I have no clue if the actual logic is correct or not, so I can only assume... but the overall integration seems like the right direction.

Perfect, that's all I was after here -- basically was looking to make sure that the vibe of the vibe was right for Lucky 🍀 .

> My only concern is that compile-time flag. I totally get it, I'm just not sure if it'll actually work. 

Will verify for sure, I'm not entirely certain either. I'm not sure it's necessary at all either -- the rest of this will be picked up and edited by me before I submit anything but when you said the prior approach wasn't very "Lucky-like" I both got it and didn't get it, so wanted to see if this was kind of more what you had in mind, in terms of structure/semantics/dev UX.

Will holler when I've got something w/more substance to review in full! 🤞

**Comment by jwoertink** [comment]
wow, that's exciting, and even better is that it only requires adding the shard and a simple require statement to enable!
